metricproviders/cloudwatch/cloudwatch.go

```diff
@@ -0,0 +1,131 @@
+package cloudwatch
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatch"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatch/types"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/evaluate"
+	metricutil "github.com/argoproj/argo-rollouts/utils/metric"
+)
+
+const (
+	//ProviderType indicates the provider is cloudwatch
+	ProviderType = "CloudWatch"
+)
+
+type CloudWatchClientAPI interface {
+	Query(interval time.Duration, query []types.MetricDataQuery) (*cloudwatch.GetMetricDataOutput, error)
+}
+
+type CloudWatchClient struct {
+	client *cloudwatch.Client
+}
+
+func (c *CloudWatchClient) Query(interval time.Duration, query []types.MetricDataQuery) (*cloudwatch.GetMetricDataOutput, error) {
+	endTime := time.Now()
+	startTime := endTime.Add(-interval)
+	return c.client.GetMetricData(context.TODO(), &cloudwatch.GetMetricDataInput{
+		StartTime:         &startTime,
+		EndTime:           &endTime,
+		MetricDataQueries: query,
+	})
+}
+
+// Provider contains all the required components to run a CloudWatch query
+// Implements the Provider Interface
+type Provider struct {
+	api    CloudWatchClientAPI
+	logCtx log.Entry
+}
+
+func (p *Provider) Type() string {
+	return ProviderType
+}
+
+// Run queries with CloudWatch provider for the metric
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	startTime := metav1.Now()
+	measurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+		Metadata:  map[string]string{},
+	}
+
+	interval := time.Minute * 5
+	if metric.Provider.CloudWatch.Interval != "" {
+		d, err := metric.Provider.CloudWatch.Interval.Duration()
+		if err != nil {
+			return metricutil.MarkMeasurementError(measurement, err)
+		}
+		interval = d
+	}
+
+	query := []types.MetricDataQuery{}
+	if err := json.Unmarshal([]byte(metric.Provider.CloudWatch.MetricDataQueries), &query); err != nil {
+		return metricutil.MarkMeasurementError(measurement, err)
+	}
+
+	result, err := p.api.Query(interval, query)
+	if err != nil {
+		return metricutil.MarkMeasurementError(measurement, err)
+	}
+
+	for _, message := range result.Messages {
+		p.logCtx.Warnf("CloudWatch returned the following messages: %+v", message)
+	}
```

Comment:
Instead of or in addition to logging a warning, the `Measurement` object has a `Metadata` field, which can be used to store any information you want to surface in the AnalysisRun. The end-user will likely not have access to the controller logs, so having messages stored in the measurement metadata will help these users debug problems.